Mathlib/SetTheory/Ordinal/Veblen.lean

```diff
@@ -0,0 +1,320 @@
+  · exact isNormal_veblenWith' f h
+
+protected alias IsNormal.veblenWith := isNormal_veblenWith
+
+theorem veblenWith_veblenWith_of_lt (h : a < b) (c : Ordinal) :
+    veblenWith f a (veblenWith f b c) = veblenWith f b c := by
+  let x : {a // a < b} := ⟨a, h⟩
+  rw [veblenWith_of_ne_zero f h.bot_lt.ne',
+    derivFamily_fp (f := fun y : Set.Iio b ↦ veblenWith f y.1) (i := x)]
+  exact hf.veblenWith x
+
+theorem veblenWith_succ (o : Ordinal) : veblenWith f (Order.succ o) = deriv (veblenWith f o) := by
+  rw [deriv_eq_enumOrd (hf.veblenWith o), veblenWith_of_ne_zero f (succ_ne_zero _),
+    derivFamily_eq_enumOrd]
+  · apply congr_arg
+    ext a
+    rw [Set.mem_iInter]
+    use fun ha ↦ ha ⟨o, Order.lt_succ o⟩
+    rintro (ha : _ = _) ⟨b, hb : b < _⟩
+    obtain rfl | hb := Order.lt_succ_iff_eq_or_lt.1 hb
+    · rw [Function.mem_fixedPoints_iff, ha]
+    · rw [← ha]
+      exact veblenWith_veblenWith_of_lt hf hb _
+  · exact fun a ↦ hf.veblenWith a
+
+theorem veblenWith_right_strictMono (o : Ordinal) : StrictMono (veblenWith f o) :=
+  (hf.veblenWith o).strictMono
+
+theorem veblenWith_lt_veblenWith_iff_right : veblenWith f o a < veblenWith f o b ↔ a < b :=
+  (veblenWith_right_strictMono hf o).lt_iff_lt
+
+theorem veblenWith_le_veblenWith_iff_right : veblenWith f o a ≤ veblenWith f o b ↔ a ≤ b :=
+  (veblenWith_right_strictMono hf o).le_iff_le
+
+theorem veblenWith_injective (o : Ordinal) : Function.Injective (veblenWith f o) :=
+  (veblenWith_right_strictMono hf o).injective
+
+theorem veblenWith_inj : veblenWith f o a = veblenWith f o b ↔ a = b :=
+  (veblenWith_injective hf o).eq_iff
+
+theorem right_le_veblenWith (a b : Ordinal) : b ≤ veblenWith f a b :=
+  (veblenWith_right_strictMono hf a).le_apply
+
+theorem veblenWith_left_monotone (o : Ordinal) : Monotone (veblenWith f · o) := by
+  rw [monotone_iff_forall_lt]
+  intro a b h
+  rw [← veblenWith_veblenWith_of_lt hf h]
+  exact (veblenWith_right_strictMono hf a).monotone (right_le_veblenWith hf b o)
+
+theorem veblenWith_pos (hp : 0 < f 0) (a b : Ordinal) : 0 < veblenWith f a b := by
+  have H (b) : 0 < veblenWith f 0 b := by
+    rw [veblenWith_zero]
+    exact hp.trans_le (hf.monotone (Ordinal.zero_le _))
+  obtain rfl | h := Ordinal.eq_zero_or_pos a
+  · exact H b
+  · rw [← veblenWith_veblenWith_of_lt hf h]
+    exact H _
+
+theorem veblenWith_zero_strictMono (hp : 0 < f 0) : StrictMono (veblenWith f · 0) := by
+  intro a b h
+  dsimp only
+  rw [← veblenWith_veblenWith_of_lt hf h, veblenWith_lt_veblenWith_iff_right hf]
+  exact veblenWith_pos hf hp b 0
+
+theorem veblenWith_zero_lt_veblenWith_zero (hp : 0 < f 0) :
+    veblenWith f a 0 < veblenWith f b 0 ↔ a < b :=
+  (veblenWith_zero_strictMono hf hp).lt_iff_lt
+
+theorem veblenWith_zero_le_veblenWith_zero (hp : 0 < f 0) :
+    veblenWith f a 0 ≤ veblenWith f b 0 ↔ a ≤ b :=
+  (veblenWith_zero_strictMono hf hp).le_iff_le
+
+theorem veblenWith_zero_inj (hp : 0 < f 0) : veblenWith f a 0 = veblenWith f b 0 ↔ a = b :=
```

Comment:
Can you please avoid using `a` and `b` for both the second arguments and third arguments? It's quite confusing (cf Bhavik and I being confused)
```suggestion
theorem veblenWith_zero_inj (hp : 0 < f 0) : veblenWith f o₁ 0 = veblenWith f o₂ 0 ↔ o₁= o₂ :=
```

Reply:
Same for other lemmas in this PR

Reply:
In truth the different arguments to the Veblen function behave somewhat similarly at a high level, as is perhaps best illustrated by the existence of a Veblen function with [transfinitely many variables](https://en.wikipedia.org/wiki/Veblen_function#Transfinitely_many_variables) `(Ordinal →₀ Ordinal) → Ordinal`. But I guess that sticking to `o` for the first variable and `a`, `b` for the second can't really hurt.